### PR TITLE
[Feature] support load model/trainer states from s3

### PR DESCRIPTION
**Pull request by eric-haibin-lin**
## Description ##
Extending #700 with functionality to pull model parameter/trainer states from s3.

### Essentials ###
- [x] PR's title starts with a category (e.g. [BUGFIX], [MODEL], [TUTORIAL], [FEATURE], [DOC], etc)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by mli** [comment]
Job PR-702/1 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-702/1/index.html

**Comment by mli** [comment]
Job PR-702/2 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-702/2/index.html

**Comment by mli** [comment]
Job PR-702/3 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-702/3/index.html

File: scripts/bert/pretraining_utils.py [hunk 1]
```diff
@@ -126,13 +127,16 @@ def __iter__(self):
 
     return DummyIter(data_batch)
 
-def save_params(step_num, model, trainer, ckpt_dir):
+def save_params(step_num, model, ckpt_dir):
```
Comment (szha) [inline scripts/bert/pretraining_utils.py]: this name can cause confusion as we advertise that people should use save_parameters.

**Comment by mli** [comment]
Job PR-702/4 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-702/4/index.html

File: src/gluonnlp/utils/parameter.py [hunk 2]
```diff
@@ -172,14 +224,20 @@ def save_states(trainer, fname):
...
-            os.makedirs(temp_dir)
-        temp_path = os.path.join(temp_dir, str(hash(os.times())))
-        trainer.save_states(temp_path)
-        _upload_file(temp_path, fname)
-        os.remove(temp_path)
-    else:
-        trainer.save_states(fname)
+    _s3_compatible_save_load(True, trainer.save_states, fname)
+
+class TempFilePath(object):
```
Comment (szha) [inline src/gluonnlp/utils/parameter.py]: it seems weird to have temp file path in parameter.py. Maybe either create a new module file in utils or make this class internal with the `_` prefix

**Comment by mli** [comment]
Job PR-702/7 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-702/7/index.html

**Comment by mli** [comment]
Job PR-702/9 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-702/9/index.html

**Pull request merged**